python2-pybind11/challenge/containers.cpp

```diff
@@ -0,0 +1,62 @@
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include "basics.hpp"
+#include "containers.hpp"
+
+namespace py = pybind11;
+
+PYBIND11_DECLARE_HOLDER_TYPE(T, std::shared_ptr<T>);
+
+namespace containers {
+
+class DoodadSetIterator {
+public:
+    /* Keep copy of DoodadSet elements (as shared_ptrs)
+     * This is needed because the iterator might outlive the container.
+     * A more elegant solution would be to have DoodadSet inherit from
+     * std::enable_shared_from_this<DoodadSet>, so that an
+     * iterator can keep the container alive while it is still in use
+     * even though the container has been garbage collected on the Python
+     * side, but that would require modifying the C++ implementation. */
```

Comment:
If the `py::return_value_policy::reference_internal` works anything like the Boost.Python internal reference policy, I suspect it can also be used to solve this problem: it'd let you put a *Python* reference to the `DoodadSet` in the Python-side `DoodadSetIterator`.